# Software Carpentry with Python: Part 2

## Data wrangling with the pandas library

For November 21, 2019

Data needs to be downloaded at:
https://go.gwu.edu/pythondata

https://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip


### Starting in the same spot:
Remember, we're in the python-lesson directory. 

1. Let's create a New > Folder. Click the checkbox and Rename it: data. 
2. Then click to go into it. 
2. We need to put our gapminder data here. You should have already downloaded that file as part of the set-up. If not, go to https://go.gwu.edu/gapminder and download it now. Unzip it!
3. Click Upload and upload the unzipped data file. 

We're setting things up this was so that we all have the same file structure and can follow along. Also, this is generally a good practice, to create a data folder and put your original data files in there. 

A quick aside that there are Python libraries like OS Library that can work with our directory structure, however, that is not our focus today.

Lessons used: Software Carpentry: https://swcarpentry.github.io/python-novice-gapminder/08-data-frames/index.html


## Why use Python for data analysis? 
* We can automate the process of performing data manipulations in Python. 
* It’s efficient to spend time building the code to perform these tasks because once it’s built, we can use it over and over on different datasets that use a similar format. 
* This makes our methods easily reproducible. We can also easily share our code with colleagues and they can replicate the same analysis.

## Working With Pandas DataFrames
One of the best options for working with tabular data in Python is to use the Python library pandas. The pandas library provides data structures, produces high quality plots with matplotlib and integrates nicely with other libraries that use NumPy (which is another Python library) arrays.

**Python doesn’t load all of the libraries available to it by default.** We have to add an import statement to our code in order to use library functions. To import a library, we use the syntax import libraryName. 

If we want to give the library a nickname to shorten the command, we can add **as nickNameHere**. An example of importing the pandas library using the common nickname pd is below.

In [216]:
import pandas as pd

### Reading CSV Data Using Pandas
We will begin by locating and reading our data which are in CSV format. CSV stands for Comma-Separated Values and is a common way store formatted data. 

We can use Pandas’ read_csv function to pull the file directly into a DataFrame.

In [217]:
pd.read_csv('data/gapminder_all.csv')

continent         country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0      Africa         Algeria     2449.008185     3013.976023     2550.816880   
1      Africa          Angola     3520.610273     3827.940465     4269.276742   
2      Africa           Benin     1062.752200      959.601080      949.499064   
3      Africa        Botswana      851.241141      918.232535      983.653976   
4      Africa    Burkina Faso      543.255241      617.183465      722.512021   
..        ...             ...             ...             ...             ...   
137    Europe     Switzerland    14734.232750    17909.489730    20431.092700   
138    Europe          Turkey     1969.100980     2218.754257     2322.869908   
139    Europe  United Kingdom     9979.508487    11283.177950    12477.177070   
140   Oceania       Australia    10039.595640    10949.649590    12217.226860   
141   Oceania     New Zealand    10556.575660    12247.395320    13175.678000   

     gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0       3246.991771     4182.663766     4910.416756     5745.160213   
1       5522.776375     5473.288005     3008.647355     2756.953672   
2       1035.831411     1085.796879     1029.161251     1277.897616   
3       1214.709294     2263.611114     3214.857818     4551.142150   
4        794.826560      854.735976      743.387037      807.198586   
..              ...             ...             ...             ...   
137    22966.144320    27195.113040    26982.290520    28397.715120   
138     2826.356387     3450.696380     4269.122326     4241.356344   
139    14142.850890    15895.116410    17428.748460    18232.424520   
140    14526.124650    16788.629480    18334.197510    19477.009280   
141    14463.918930    16046.037280    16233.717700    17632.410400   

     gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972    pop_1977  \
0       5681.358539  ...  11000948.0  12760499.0  14760787.0  17152804.0   
1       2430.208311  ...   4826015.0   5247469.0   5894858.0   6162675.0   
2       1225.856010  ...   2151895.0   2427334.0   2761407.0   3168267.0   
3       6205.883850  ...    512764.0    553541.0    619351.0    781472.0   
4        912.063142  ...   4919632.0   5127935.0   5433886.0   5889574.0   
..              ...  ...         ...         ...         ...         ...   
137    30281.704590  ...   5666000.0   6063000.0   6401400.0   6316424.0   
138     5089.043686  ...  29788695.0  33411317.0  37492953.0  42404033.0   
139    21664.787670  ...  53292000.0  54959000.0  56079000.0  56179000.0   
140    21888.889030  ...  10794968.0  11872264.0  13177000.0  14074100.0   
141    19007.191290  ...   2488550.0   2728150.0   2929100.0   3164900.0   

       pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  
0    20033753.0  23254956.0  26298373.0  29072015.0  31287142  33333216  
1     7016384.0   7874230.0   8735988.0   9875024.0  10866106  12420476  
2     3641603.0   4243788.0   4981671.0   6066080.0   7026113   8078314  
3      970347.0   1151184.0   1342614.0   1536536.0   1630347   1639131  
4     6634596.0   7586551.0   8878303.0  10352843.0  12251209  14326203  
..          ...         ...         ...         ...       ...       ...  
137   6468126.0   6649942.0   6995447.0   7193761.0   7361757   7554661  
138  47328791.0  52881328.0  58179144.0  63047647.0  67308928  71158647  
139  56339704.0  56981620.0  57866349.0  58808266.0  59912431  60776238  
140  15184200.0  16257249.0  17481977.0  18565243.0  19546792  20434176  
141   3210650.0   3317166.0   3437674.0   3676187.0   3908037   4115771  

[142 rows x 38 columns]

This output is the rows in our CSV file, now as a pandas DataFrame object. 

The first column is the index of the DataFrame. The index is used to identify the position of the data, but it is not an actual column of the DataFrame. 

It looks like the read_csv function in Pandas read our file properly. However, we haven’t saved any data to memory so we can work with it. We need to assign the DataFrame to a variable. Remember that a variable is a name for a value, such as x, or data. We can create a new object with a variable name by assigning a value to it using =.

In [218]:
data = pd.read_csv("data/gapminder_all.csv", index_col="country")

There are many ways to summarize and access the data stored in DataFrames, using attributes and methods provided by the DataFrame object.

Methods are called on a DataFrame object using the syntax df_object.method(). As an example, `data.head()` gets the first few rows in the DataFrame surveys_df using the head() method. With a method, we can supply extra information in the parens to control behaviour.

In [219]:
# nothing in parens defaults to first 5 rows
data.head()

continent  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                                  
Algeria         Africa     2449.008185     3013.976023     2550.816880   
Angola          Africa     3520.610273     3827.940465     4269.276742   
Benin           Africa     1062.752200      959.601080      949.499064   
Botswana        Africa      851.241141      918.232535      983.653976   
Burkina Faso    Africa      543.255241      617.183465      722.512021   

              gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
country                                                                        
Algeria          3246.991771     4182.663766     4910.416756     5745.160213   
Angola           5522.776375     5473.288005     3008.647355     2756.953672   
Benin            1035.831411     1085.796879     1029.161251     1277.897616   
Botswana         1214.709294     2263.611114     3214.857818     4551.142150   
Burkina Faso      794.826560      854.735976      743.387037      807.198586   

              gdpPercap_1987  gdpPercap_1992  ...    pop_1962    pop_1967  \
country                                       ...                           
Algeria          5681.358539     5023.216647  ...  11000948.0  12760499.0   
Angola           2430.208311     2627.845685  ...   4826015.0   5247469.0   
Benin            1225.856010     1191.207681  ...   2151895.0   2427334.0   
Botswana         6205.883850     7954.111645  ...    512764.0    553541.0   
Burkina Faso      912.063142      931.752773  ...   4919632.0   5127935.0   

                pop_1972    pop_1977    pop_1982    pop_1987    pop_1992  \
country                                                                    
Algeria       14760787.0  17152804.0  20033753.0  23254956.0  26298373.0   
Angola         5894858.0   6162675.0   7016384.0   7874230.0   8735988.0   
Benin          2761407.0   3168267.0   3641603.0   4243788.0   4981671.0   
Botswana        619351.0    781472.0    970347.0   1151184.0   1342614.0   
Burkina Faso   5433886.0   5889574.0   6634596.0   7586551.0   8878303.0   

                pop_1997  pop_2002  pop_2007  
country                                       
Algeria       29072015.0  31287142  33333216  
Angola         9875024.0  10866106  12420476  
Benin          6066080.0   7026113   8078314  
Botswana       1536536.0   1630347   1639131  
Burkina Faso  10352843.0  12251209  14326203  

[5 rows x 37 columns]

In [220]:
type(data)

pandas.core.frame.DataFrame

In [221]:
data.dtypes

continent          object
gdpPercap_1952    float64
gdpPercap_1957    float64
gdpPercap_1962    float64
gdpPercap_1967    float64
gdpPercap_1972    float64
gdpPercap_1977    float64
gdpPercap_1982    float64
gdpPercap_1987    float64
gdpPercap_1992    float64
gdpPercap_1997    float64
gdpPercap_2002    float64
gdpPercap_2007    float64
lifeExp_1952      float64
lifeExp_1957      float64
lifeExp_1962      float64
lifeExp_1967      float64
lifeExp_1972      float64
lifeExp_1977      float64
lifeExp_1982      float64
lifeExp_1987      float64
lifeExp_1992      float64
lifeExp_1997      float64
lifeExp_2002      float64
lifeExp_2007      float64
pop_1952          float64
pop_1957          float64
pop_1962          float64
pop_1967          float64
pop_1972          float64
pop_1977          float64
pop_1982          float64
pop_1987          float64
pop_1992          float64
pop_1997          float64
pop_2002            int64
pop_2007            int64
dtype: object

The DataFrame.columns variable stores information about the dataframe’s columns. 

Note that this is data, not a method. Like math.pi. So do not use () to try to call it.

In [222]:
data.columns

Index(['continent', 'gdpPercap_1952', 'gdpPercap_1957', 'gdpPercap_1962',
       'gdpPercap_1967', 'gdpPercap_1972', 'gdpPercap_1977', 'gdpPercap_1982',
       'gdpPercap_1987', 'gdpPercap_1992', 'gdpPercap_1997', 'gdpPercap_2002',
       'gdpPercap_2007', 'lifeExp_1952', 'lifeExp_1957', 'lifeExp_1962',
       'lifeExp_1967', 'lifeExp_1972', 'lifeExp_1977', 'lifeExp_1982',
       'lifeExp_1987', 'lifeExp_1992', 'lifeExp_1997', 'lifeExp_2002',
       'lifeExp_2007', 'pop_1952', 'pop_1957', 'pop_1962', 'pop_1967',
       'pop_1972', 'pop_1977', 'pop_1982', 'pop_1987', 'pop_1992', 'pop_1997',
       'pop_2002', 'pop_2007'],
      dtype='object')

In [223]:
data.describe()

gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
count      142.000000      142.000000      142.000000      142.000000   
mean      3725.276046     4299.408345     4725.812342     5483.653047   
std       9321.064786     9869.662202     8667.362525     8095.315431   
min        298.846212      335.997115      355.203227      349.000000   
25%        864.752389      930.540819     1059.149171     1151.245103   
50%       1968.528344     2173.220291     2335.439533     2678.334741   
75%       3913.492777     4876.356362     5709.381428     7075.932943   
max     108382.352900   113523.132900    95458.111760    80894.883260   

       gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
count      142.000000      142.000000      142.000000      142.000000   
mean      6770.082815     7313.166421     7518.901673     7900.920218   
std      10614.383403     8362.489150     7733.845006     8288.281304   
min        357.000000      371.000000      424.000000      385.000000   
25%       1257.193853     1357.257252     1363.338985     1327.469823   
50%       3339.129407     3798.609244     4216.228428     4280.300366   
75%       9508.839303    11204.102423    12347.953723    11994.052795   
max     109347.867000    59265.477140    33693.175250    31540.974800   

       gdpPercap_1992  gdpPercap_1997  ...      pop_1962      pop_1967  \
count      142.000000      142.000000  ...  1.420000e+02  1.420000e+02   
mean      8158.608521     9090.175363  ...  2.042101e+07  2.265830e+07   
std       9031.846080    10171.493263  ...  6.978865e+07  7.837548e+07   
min        347.000000      312.188423  ...  6.534500e+04  7.078700e+04   
25%       1270.660958     1366.837958  ...  1.784362e+06  2.034768e+06   
50%       4386.085502     4781.825478  ...  4.686040e+06  5.170176e+06   
75%      10684.351870    12022.867188  ...  1.098008e+07  1.261458e+07   
max      34932.919590    41283.164330  ...  6.657700e+08  7.545500e+08   

           pop_1972      pop_1977      pop_1982      pop_1987      pop_1992  \
count  1.420000e+02  1.420000e+02  1.420000e+02  1.420000e+02  1.420000e+02   
mean   2.518998e+07  2.767638e+07  3.020730e+07  3.303857e+07  3.599092e+07   
std    8.864682e+07  9.748109e+07  1.050986e+08  1.147562e+08  1.245026e+08   
min    7.659500e+04  8.679600e+04  9.859300e+04  1.108120e+05  1.259110e+05   
25%    2.351192e+06  2.759717e+06  3.006286e+06  3.194990e+06  3.605992e+06   
50%    5.877996e+06  6.404036e+06  7.007320e+06  7.774862e+06  8.688686e+06   
75%    1.467920e+07  1.667023e+07  1.840732e+07  2.094754e+07  2.270538e+07   
max    8.620300e+08  9.434550e+08  1.000281e+09  1.084035e+09  1.164970e+09   

           pop_1997      pop_2002      pop_2007  
count  1.420000e+02  1.420000e+02  1.420000e+02  
mean   3.883947e+07  4.145759e+07  4.402122e+07  
std    1.334174e+08  1.408483e+08  1.476214e+08  
min    1.456080e+05  1.703720e+05  1.995790e+05  
25%    3.770150e+06  4.173506e+06  4.508034e+06  
50%    9.735064e+06  1.037292e+07  1.051753e+07  
75%    2.431137e+07  2.654556e+07  3.121004e+07  
max    1.230075e+09  1.280400e+09  1.318683e+09  

[8 rows x 36 columns]

## Indexing and Slicing in Python
We often want to work with subsets of a DataFrame object. There are different ways to accomplish this including: using:
* labels (column headings)
* numeric ranges
* specific x,y index locations.

### Selecting data using Labels (Column Headings)
We use square brackets [] to select a subset of a Python object. For example, we can select all data from a column named lifeExp2007 from the data DataFrame by name. There are two ways to do this:

In [30]:
# Method 1: select a 'subset' of the data using the column name
data['lifeExp_2007']

country
Algeria           72.301
Angola            42.731
Benin             56.728
Botswana          50.728
Burkina Faso      52.295
                   ...  
Switzerland       81.701
Turkey            71.777
United Kingdom    79.425
Australia         81.235
New Zealand       80.204
Name: lifeExp_2007, Length: 142, dtype: float64

In [32]:
# Method 2: use the column name as an 'attribute'; gives the same output
data.lifeExp_2007

country
Algeria           72.301
Angola            42.731
Benin             56.728
Botswana          50.728
Burkina Faso      52.295
                   ...  
Switzerland       81.701
Turkey            71.777
United Kingdom    79.425
Australia         81.235
New Zealand       80.204
Name: lifeExp_2007, Length: 142, dtype: float64

A DataFrame is a collection of Series; The DataFrame is the way Pandas represents a table, and Series is the data-structure Pandas use to represent a column.

What we did above by taking a column was creating a Series.

Pandas is built on top of the Numpy library, which in practice means that most of the methods defined for Numpy Arrays apply to Pandas Series/DataFrames.

What makes Pandas so attractive is the powerful interface to access individual records of the table, proper handling of missing values, and relational-databases operations between DataFrames.


### Selecting using slices of rows 

Slicing using the `[]` operator selects a set of rows and/or columns from a DataFrame, not counting the labels. To slice out a set of rows, you use the following syntax: data[start:stop]. When slicing in pandas the start bound is included in the output. The stop bound is one step BEYOND the row you want to select. So if you want to select rows 0, 1 and 2 your code would look like this:

In [62]:
data[0:3]

continent  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                             
Algeria    Africa     2449.008185     3013.976023     2550.816880   
Angola     Africa     3520.610273     3827.940465     4269.276742   
Benin      Africa     1062.752200      959.601080      949.499064   

         gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
country                                                                   
Algeria     3246.991771     4182.663766     4910.416756     5745.160213   
Angola      5522.776375     5473.288005     3008.647355     2756.953672   
Benin       1035.831411     1085.796879     1029.161251     1277.897616   

         gdpPercap_1987  gdpPercap_1992  ...    pop_1962    pop_1967  \
country                                  ...                           
Algeria     5681.358539     5023.216647  ...  11000948.0  12760499.0   
Angola      2430.208311     2627.845685  ...   4826015.0   5247469.0   
Benin       1225.856010     1191.207681  ...   2151895.0   2427334.0   

           pop_1972    pop_1977    pop_1982    pop_1987    pop_1992  \
country                                                               
Algeria  14760787.0  17152804.0  20033753.0  23254956.0  26298373.0   
Angola    5894858.0   6162675.0   7016384.0   7874230.0   8735988.0   
Benin     2761407.0   3168267.0   3641603.0   4243788.0   4981671.0   

           pop_1997  pop_2002  pop_2007  
country                                  
Algeria  29072015.0  31287142  33333216  
Angola    9875024.0  10866106  12420476  
Benin     6066080.0   7026113   8078314  

[3 rows x 37 columns]

In [64]:
# Can also leave out the 0
data[:3]

continent  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                             
Algeria    Africa     2449.008185     3013.976023     2550.816880   
Angola     Africa     3520.610273     3827.940465     4269.276742   
Benin      Africa     1062.752200      959.601080      949.499064   

         gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
country                                                                   
Algeria     3246.991771     4182.663766     4910.416756     5745.160213   
Angola      5522.776375     5473.288005     3008.647355     2756.953672   
Benin       1035.831411     1085.796879     1029.161251     1277.897616   

         gdpPercap_1987  gdpPercap_1992  ...    pop_1962    pop_1967  \
country                                  ...                           
Algeria     5681.358539     5023.216647  ...  11000948.0  12760499.0   
Angola      2430.208311     2627.845685  ...   4826015.0   5247469.0   
Benin       1225.856010     1191.207681  ...   2151895.0   2427334.0   

           pop_1972    pop_1977    pop_1982    pop_1987    pop_1992  \
country                                                               
Algeria  14760787.0  17152804.0  20033753.0  23254956.0  26298373.0   
Angola    5894858.0   6162675.0   7016384.0   7874230.0   8735988.0   
Benin     2761407.0   3168267.0   3641603.0   4243788.0   4981671.0   

           pop_1997  pop_2002  pop_2007  
country                                  
Algeria  29072015.0  31287142  33333216  
Angola    9875024.0  10866106  12420476  
Benin     6066080.0   7026113   8078314  

[3 rows x 37 columns]

## Selecting values

To access a value at the position row i, column j [i,j] of a DataFrame, we have two options, depending on what is the meaning of i in use. Remember that a DataFrame provides a index as a way to identify the rows of the table; a row, then, has a position inside the table as well as a label, which uniquely identifies its entry in the DataFrame.

Use DataFrame.loc[..., ...] to select values by their label.

Use `: ` on its own to mean all columns or all rows

In [68]:
data.loc["Algeria", "gdpPercap_1952"]

2449.008185

Use `DataFrame.iloc[..., ...]` to select values by their (entry) position

Can specify location by numerical index analogously to 2D version of character selection in strings. (Or items in lists). 

The labels aren't included in the counting, they apply to the data.

In [48]:
data.iloc[0,0]

'Africa'

Let's do the same thing to get at the gdp in 1952 for Algeria:

In [69]:
data.iloc[0, 1]

2449.008185

**Exercise:**

Practice Series and slicing.

1) Get the lifeExp for all countries in 1992 and assign it to a variable called lifeExp1992. 

2) Get the GDP in 1962 for New Zealand using multiple methods of slicing. Remember we can use .tail() to see the end of the data.


**Answer #1**

In [52]:
lifeExp1992 = data["lifeExp_1992"]
print(lifeExp1992)

country
Algeria           67.744
Angola            40.647
Benin             53.919
Botswana          62.745
Burkina Faso      50.260
                   ...  
Switzerland       78.030
Turkey            66.146
United Kingdom    76.420
Australia         77.560
New Zealand       76.330
Name: lifeExp_1992, Length: 142, dtype: float64


**Answer #2**

In [84]:
data.loc["New Zealand", "gdpPercap_1962"]

13175.678

In [257]:
#There are 142 rows in our dataframe, however, since we count starting with zero, the last row is 141. 
data.iloc[141,3]

# answer is 13175.678000

13175.678

In [258]:
#help(pd.DataFrame.loc)

**Using labels on multiple rows and columns:**

To slice the life expectancy from 1992, 2002, 2007, for all of the countries:

Specify the rows we want and then the columns we want. 

In [209]:
data.loc[:,["lifeExp_1992","lifeExp_2002","lifeExp_2007"]]

lifeExp_1992  lifeExp_2002  lifeExp_2007
country                                                 
Algeria               67.744        70.994        72.301
Angola                40.647        41.003        42.731
Benin                 53.919        54.406        56.728
Botswana              62.745        46.634        50.728
Burkina Faso          50.260        50.650        52.295
...                      ...           ...           ...
Switzerland           78.030        80.620        81.701
Turkey                66.146        70.845        71.777
United Kingdom        76.420        78.471        79.425
Australia             77.560        80.370        81.235
New Zealand           76.330        79.110        80.204

[142 rows x 3 columns]

In [87]:
# A list of the labels of the columns we want.

data.loc[["Benin", "Turkey", "Afghanistan"],["lifeExp_1992","lifeExp_2002","lifeExp_2007"]]

lifeExp_1992  lifeExp_2002  lifeExp_2007
country                                              
Benin              53.919        54.406        56.728
Turkey             66.146        70.845        71.777
Afghanistan        41.674        42.129        43.828

Creating slices lets us then use methods on those subsets. For example, we earlier grabbed just the column that had the lifeExp for 1992. We can find the max and min for that slice as follows:


In [86]:
lifeExp1992 = data["lifeExp_1992"]
print(lifeExp1992.min())
print(lifeExp1992.max())

23.599
79.36


23.599 is Rwanda
79.36 is Japan

When pandas selects a single column from a DataFrame, pandas creates a view and not a copy. A view just means that no new object has been created. No new object is created, just a new reference to the one already in existence. Since no new data has been created, the assignment will modify the original DataFrame.

### Subsetting Data Using Criteria

We can also select a subset of our data using criteria. 

In [119]:
data[data.lifeExp_2007 > 80]

continent  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                                     
Canada           Americas    11367.161120    12489.950060    13462.485550   
Hong Kong China      Asia     3054.421209     3629.076457     4692.648272   
Israel               Asia     4086.522128     5385.278451     7105.630706   
Japan                Asia     3216.956347     4317.694365     6576.649461   
France             Europe     7029.809327     8662.834898    10560.485530   
Iceland            Europe     7267.688428     9244.001412    10350.159060   
Italy              Europe     4931.404155     6248.656232     8243.582340   
Norway             Europe    10095.421720    11653.973040    13450.401510   
Spain              Europe     3834.034742     4564.802410     5693.843879   
Sweden             Europe     8527.844662     9911.878226    12329.441920   
Switzerland        Europe    14734.232750    17909.489730    20431.092700   
Australia         Oceania    10039.595640    10949.649590    12217.226860   
New Zealand       Oceania    10556.575660    12247.395320    13175.678000   

                 gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
country                                                           
Canada             16076.588030    18970.570860     22090.88306   
Hong Kong China     6197.962814     8315.928145     11186.14125   
Israel              8393.741404    12786.932230     13306.61921   
Japan               9847.788607    14778.786360     16610.37701   
France             12999.917660    16107.191710     18292.63514   
Iceland            13319.895680    15798.063620     19654.96247   
Italy              10022.401310    12269.273780     14255.98475   
Norway             16361.876470    18965.055510     23311.34939   
Spain               7993.512294    10638.751310     13236.92117   
Sweden             15258.296970    17832.024640     18855.72521   
Switzerland        22966.144320    27195.113040     26982.29052   
Australia          14526.124650    16788.629480     18334.19751   
New Zealand        14463.918930    16046.037280     16233.71770   

                 gdpPercap_1982  gdpPercap_1987  gdpPercap_1992  ...  \
country                                                          ...   
Canada              22898.79214     26626.51503     26342.88426  ...   
Hong Kong China     14560.53051     20038.47269     24757.60301  ...   
Israel              15367.02920     17122.47986     18051.52254  ...   
Japan               19384.10571     22375.94189     26824.89511  ...   
France              20293.89746     22066.44214     24703.79615  ...   
Iceland             23269.60750     26923.20628     25144.39201  ...   
Italy               16537.48350     19207.23482     22013.64486  ...   
Norway              26298.63531     31540.97480     33965.66115  ...   
Spain               13926.16997     15764.98313     18603.06452  ...   
Sweden              20667.38125     23586.92927     23880.01683  ...   
Switzerland         28397.71512     30281.70459     31871.53030  ...   
Australia           19477.00928     21888.88903     23424.76683  ...   
New Zealand         17632.41040     19007.19129     18363.32494  ...   

                   pop_1962     pop_1967     pop_1972     pop_1977  \
country                                                              
Canada           18985849.0   20819767.0   22284500.0   23796400.0   
Hong Kong China   3305200.0    3722800.0    4115700.0    4583700.0   
Israel            2310904.0    2693585.0    3095893.0    3495918.0   
Japan            95831757.0  100825279.0  107188273.0  113872473.0   
France           47124000.0   49569000.0   51732000.0   53165019.0   
Iceland            182053.0     198676.0     209275.0     221823.0   
Italy            50843200.0   52667100.0   54365564.0   56059245.0   
Norway            3638919.0    3786019.0    3933004.0    4043205.0   
Spain            31158061.0   32850275.0   34513161.0   36439000.0   
Swe

Quick review of conditions: `==  != > < >= <= `

**Exercise 5**

Create a subset of the data that contains rows for countries where the GDP per capita in 2007 was less than $1000

In [254]:
data[data.gdpPercap_2007 <= 1000]


continent  gdpPercap_1952  gdpPercap_1957  \
country                                                              
Burundi                     Africa      339.296459      379.564628   
Central African Republic    Africa     1071.310713     1190.844328   
Comoros                     Africa     1102.990936     1211.148548   
Congo Dem. Rep.             Africa      780.542326      905.860230   
Eritrea                     Africa      328.940557      344.161886   
Ethiopia                    Africa      362.146280      378.904163   
Gambia                      Africa      485.230659      520.926711   
Guinea                      Africa      510.196492      576.267025   
Guinea-Bissau               Africa      299.850319      431.790457   
Liberia                     Africa      575.572996      620.969990   
Malawi                      Africa      369.165080      416.369806   
Mozambique                  Africa      468.526038      495.586833   
Niger                       Africa      761.879376      835.523403   
Rwanda                      Africa      493.323875      540.289398   
Sierra Leone                Africa      879.787736     1004.484437   
Somalia                     Africa     1135.749842     1258.147413   
Togo                        Africa      859.808657      925.908320   
Zimbabwe                    Africa      406.884115      518.764268   
Afghanistan                   Asia      779.445314      820.853030   
Myanmar                       Asia      331.000000      350.000000   

                          gdpPercap_1962  gdpPercap_1967  gdpPercap_1972  \
country                                                                    
Burundi                       355.203227      412.977514      464.099504   
Central African Republic     1193.068753     1136.056615     1070.013275   
Comoros                      1406.648278     1876.029643     1937.577675   
Congo Dem. Rep.               896.314634      861.593242      904.896069   
Eritrea                       380.995843      468.794970      514.324208   
Ethiopia                      419.456416      516.118644      566.243944   
Gambia                        599.650276      734.782912      756.086836   
Guinea                        686.373674      708.759541      741.666231   
Guinea-Bissau                 522.034373      715.580640      820.224588   
Liberia                       634.195163      713.603648      803.005454   
Malawi                        427.901086      495.514781      584.621971   
Mozambique                    556.686354      566.669154      724.917804   
Niger                         997.766113     1054.384891      954.209236   
Rwanda                        597.473073      510.963714      590.580664   
Sierra Leone                 1116.639877     1206.043465     1353.759762   
Somalia                      1369.488336     1284.733180     1254.576127   
Togo                         1067.534810     1477.596760     1649.660188   
Zimbabwe                      527.272182      569.795071      799.362176   
Afghanistan                   853.100710      836.197138      739.981106   
Myanmar                       388.000000      349.000000      357.000000   

                          gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                    
Burundi                       556.103265      559.603231      621.818819   
Central African Republic     1109.374338      956.752991      844.876350   
Comoros                      1172.603047     1267.100083     1315.980812   
Congo Dem. Rep.               795.757282      673.747818      672.774812   
Eritrea                       505.753808      524.875849      521.134133   
Ethiopia                      556.808383      577.860747      573.741314   
Gambia                        884.755251      835.809611      611.658861   
Guinea                        874.685864      857.250358      805.572472   
Guinea-Bissau                 764.725963      838.1

### Using a mask to identify a specific condition.

A mask can be useful to locate where a particular subset of values exist or don't exist, for example, NaN or "not a number". 

Comparison or function is applied element by element. Returns a similarly-shaped dataframe of True and False.

Boolean is a Python data type, True or False. False is Python's way of saying "No."

In [124]:
x = 1
x > 5

False

There is a way with pandas' methods to check for null values, (missing data or NaN). 

In [120]:
pd.isnull(data)

continent  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                                     
Algeria             False           False           False           False   
Angola              False           False           False           False   
Benin               False           False           False           False   
Botswana            False           False           False           False   
Burkina Faso        False           False           False           False   
...                   ...             ...             ...             ...   
Switzerland         False           False           False           False   
Turkey              False           False           False           False   
United Kingdom      False           False           False           False   
Australia           False           False           False           False   
New Zealand         False           False           False           False   

                gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
country                                                          
Algeria                  False           False           False   
Angola                   False           False           False   
Benin                    False           False           False   
Botswana                 False           False           False   
Burkina Faso             False           False           False   
...                        ...             ...             ...   
Switzerland              False           False           False   
Turkey                   False           False           False   
United Kingdom           False           False           False   
Australia                False           False           False   
New Zealand              False           False           False   

                gdpPercap_1982  gdpPercap_1987  gdpPercap_1992  ...  pop_1962  \
country                                                         ...             
Algeria                  False           False           False  ...     False   
Angola                   False           False           False  ...     False   
Benin                    False           False           False  ...     False   
Botswana                 False           False           False  ...     False   
Burkina Faso             False           False           False  ...     False   
...                        ...             ...             ...  ...       ...   
Switzerland              False           False           False  ...     False   
Turkey                   False           False           False  ...     False   
United Kingdom           False           False           False  ...     False   
Australia                False           False           False  ...     False   
New Zealand              False           False           False  ...     False   

                pop_1967  pop_1972  pop_1977  pop_1982  pop_1987  pop_1992  \
country                                                                      
Algeria            False     False     False     False     False     False   
Angola             False     False     False     False     False     False   
Benin              False     False     False     False     False     False   
Botswana           False     False     False     False     False     False   
Burkina Faso       False     False     False     False     False     False   
...                  ...       ...       ...       ...       ...       ...   
Switzerland        False     False     False     False     False     False   
Turkey             False     False     False     False     False     False   
United Kingdom     False     False     False     False     False     False   
Australia          False     False     False     False     False     False   
New Zealand        False     False     False     False     False     False   

                pop_1997  pop_2002  pop_2007  
country                                

This is very clean data, all data is here, none of the cells have a value of True, isnull? = False.

Let's confirm by taking a closer look, applying a filter to the data. 

using the any() method, which looks for only True values. 

In [145]:
data[pd.isnull(data).any(axis=1)]

Empty DataFrame
Columns: [continent, gdpPercap_1952, gdpPercap_1957, gdpPercap_1962, gdpPercap_1967, gdpPercap_1972, gdpPercap_1977, gdpPercap_1982, gdpPercap_1987, gdpPercap_1992, gdpPercap_1997, gdpPercap_2002, gdpPercap_2007, lifeExp_1952, lifeExp_1957, lifeExp_1962, lifeExp_1967, lifeExp_1972, lifeExp_1977, lifeExp_1982, lifeExp_1987, lifeExp_1992, lifeExp_1997, lifeExp_2002, lifeExp_2007, pop_1952, pop_1957, pop_1962, pop_1967, pop_1972, pop_1977, pop_1982, pop_1987, pop_1992, pop_1997, pop_2002, pop_2007]
Index: []

[0 rows x 37 columns]

## Group By: split-apply-combine
Pandas vectorizing methods and grouping operations are features that provide users much flexibility to analyse their data.

For instance, let’s say we want to have a clearer view on how the European countries split according to their GDP.

We can split the countries in two groups during the years surveyed, those who presented a GDP higher than the European average and those with a lower GDP.

Remember we can use methods like .mean() on a dataframe. .mean() is calculated per column. 

In [188]:
# first create a new DataFrame that is a subset, just those countries in the continent of Europe.
europe_df = data[data.continent == "Europe"].copy()
europe_df.head()

continent  gdpPercap_1952  gdpPercap_1957  \
country                                                            
Albania                   Europe     1601.056136     1942.284244   
Austria                   Europe     6137.076492     8842.598030   
Belgium                   Europe     8343.105127     9714.960623   
Bosnia and Herzegovina    Europe      973.533195     1353.989176   
Bulgaria                  Europe     2444.286648     3008.670727   

                        gdpPercap_1962  gdpPercap_1967  gdpPercap_1972  \
country                                                                  
Albania                    2312.888958     2760.196931     3313.422188   
Austria                   10750.721110    12834.602400    16661.625600   
Belgium                   10991.206760    13149.041190    16672.143560   
Bosnia and Herzegovina     1709.683679     2172.352423     2860.169750   
Bulgaria                   4254.337839     5577.002800     6597.494398   

                        gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                  
Albania                    3533.003910     3630.880722     3738.932735   
Austria                   19749.422300    21597.083620    23687.826070   
Belgium                   19117.974480    20979.845890    22525.563080   
Bosnia and Herzegovina     3528.481305     4126.613157     4314.114757   
Bulgaria                   7612.240438     8224.191647     8239.854824   

                        gdpPercap_1992  ...   pop_1962   pop_1967   pop_1972  \
country                                 ...                                    
Albania                    2497.437901  ...  1728137.0  1984060.0  2263554.0   
Austria                   27042.018680  ...  7129864.0  7376998.0  7544201.0   
Belgium                   25575.570690  ...  9218400.0  9556500.0  9709100.0   
Bosnia and Herzegovina     2546.781445  ...  3349000.0  3585000.0  3819000.0   
Bulgaria                   6302.623438  ...  8012946.0  8310226.0  8576200.0   

                         pop_1977   pop_1982   pop_1987    pop_1992  \
country                                                               
Albania                 2509048.0  2780097.0  3075321.0   3326498.0   
Austria                 7568430.0  7574613.0  7578903.0   7914969.0   
Belgium                 9821800.0  9856303.0  9870200.0  10045622.0   
Bosnia and Herzegovina  4086000.0  4172693.0  4338977.0   4256013.0   
Bulgaria                8797022.0  8892098.0  8971958.0   8658506.0   

                          pop_1997  pop_2002  pop_2007  
country                                                 
Albania                  3428038.0   3508512   3600523  
Austria                  8069876.0   8148312   8199783  
Belgium                 10199787.0  10311970  10392226  
Bosnia and Herzegovina   3607000.0   4165416   4552198  
Bulgaria                 8066057.0   7661799   7322858  

[5 rows x 37 columns]

Looks like we still have ALL of the columns, not just GDP, so let's further slice the df. 

Going to use `.iloc` because there are a dozen columns out of the 37 we want and I don't want to list them all. 

In [256]:
# overwrite our existing dataframe, and use iloc to get all rows, just the columns with the GDP per capita variables.
europe_df = europe_df.iloc[:,1:13]

In [190]:
europe_df.head()

gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                                  
Albania                    1601.056136     1942.284244     2312.888958   
Austria                    6137.076492     8842.598030    10750.721110   
Belgium                    8343.105127     9714.960623    10991.206760   
Bosnia and Herzegovina      973.533195     1353.989176     1709.683679   
Bulgaria                   2444.286648     3008.670727     4254.337839   

                        gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
country                                                                  
Albania                    2760.196931     3313.422188     3533.003910   
Austria                   12834.602400    16661.625600    19749.422300   
Belgium                   13149.041190    16672.143560    19117.974480   
Bosnia and Herzegovina     2172.352423     2860.169750     3528.481305   
Bulgaria                   5577.002800     6597.494398     7612.240438   

                        gdpPercap_1982  gdpPercap_1987  gdpPercap_1992  \
country                                                                  
Albania                    3630.880722     3738.932735     2497.437901   
Austria                   21597.083620    23687.826070    27042.018680   
Belgium                   20979.845890    22525.563080    25575.570690   
Bosnia and Herzegovina     4126.613157     4314.114757     2546.781445   
Bulgaria                   8224.191647     8239.854824     6302.623438   

                        gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
country                                                                 
Albania                    3193.054604     4604.211737     5937.029526  
Austria                   29095.920660    32417.607690    36126.492700  
Belgium                   27561.196630    30485.883750    33692.605080  
Bosnia and Herzegovina     4766.355904     6018.975239     7446.298803  
Bulgaria                   5970.388760     7696.777725    10680.792820

We can use mean() to get the mean of each column. 

In [211]:
europe_df.mean()

gdpPercap_1952     5661.057435
gdpPercap_1957     6963.012816
gdpPercap_1962     8365.486814
gdpPercap_1967    10143.823757
gdpPercap_1972    12479.575246
gdpPercap_1977    14283.979110
gdpPercap_1982    15617.896551
gdpPercap_1987    17214.310727
gdpPercap_1992    17061.568084
gdpPercap_1997    19076.781802
gdpPercap_2002    21711.732422
gdpPercap_2007    25054.481636
wealth_score          0.154955
dtype: float64

What is the result of taking mean() on a whole DataFrame?

In [212]:
type(europe_df.mean())

pandas.core.series.Series

Now we know the average GDP for each of the years. Next is to figure out whether each country's GDP is over the mean, by creating a boolean mask like we did earlier. Remember the mask is a DataFrame

In [262]:
mask_higher = europe_df > europe_df.mean()
mask_higher.head()

gdpPercap_1962  gdpPercap_1967  gdpPercap_1972  \
country                                                                  
Albania                          False           False           False   
Austria                           True            True            True   
Belgium                           True            True            True   
Bosnia and Herzegovina           False           False           False   
Bulgaria                         False           False           False   

                        gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                  
Albania                          False           False           False   
Austria                           True            True            True   
Belgium                           True            True            True   
Bosnia and Herzegovina           False           False           False   
Bulgaria                         False           False           False   

                        gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  \
country                                                                  
Albania                          False           False           False   
Austria                           True            True            True   
Belgium                           True            True            True   
Bosnia and Herzegovina           False           False           False   
Bulgaria                         False           False           False   

                        gdpPercap_2007  wealth_score  
country                                               
Albania                          False         False  
Austria                           True          True  
Belgium                           True          True  
Bosnia and Herzegovina           False         False  
Bulgaria                         False         False

We then estimate a wealthy score based on the historical (from 1962 to 2007) values, where we count how many times a country has participated in the groups of lower or higher GDP. So, need to count how many Trues there are in each row.

We can use the **aggregate()** method to count (or sum), and then use axis=1 because we're applying this horizontally, across the all of the columns in a row. Axis = 0 is often the default in pandas and that applies something down a column. 

In [263]:
wealth_score = mask_higher.aggregate('sum', axis=1) / len(europe_df.columns)
wealth_score.head()

country
Albania                   0.0
Austria                   1.0
Belgium                   1.0
Bosnia and Herzegovina    0.0
Bulgaria                  0.0
dtype: float64

In [264]:
type(wealth_score)

pandas.core.series.Series

We can now add this back to our dataframe since it's a pandas Series and it has the same index. (They'll be able to matched up)

In [265]:
europe_df["wealth_score"] = wealth_score
europe_df.head()

gdpPercap_1962  gdpPercap_1967  gdpPercap_1972  \
country                                                                  
Albania                    2312.888958     2760.196931     3313.422188   
Austria                   10750.721110    12834.602400    16661.625600   
Belgium                   10991.206760    13149.041190    16672.143560   
Bosnia and Herzegovina     1709.683679     2172.352423     2860.169750   
Bulgaria                   4254.337839     5577.002800     6597.494398   

                        gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                  
Albania                    3533.003910     3630.880722     3738.932735   
Austria                   19749.422300    21597.083620    23687.826070   
Belgium                   19117.974480    20979.845890    22525.563080   
Bosnia and Herzegovina     3528.481305     4126.613157     4314.114757   
Bulgaria                   7612.240438     8224.191647     8239.854824   

                        gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  \
country                                                                  
Albania                    2497.437901     3193.054604     4604.211737   
Austria                   27042.018680    29095.920660    32417.607690   
Belgium                   25575.570690    27561.196630    30485.883750   
Bosnia and Herzegovina     2546.781445     4766.355904     6018.975239   
Bulgaria                   6302.623438     5970.388760     7696.777725   

                        gdpPercap_2007  wealth_score  
country                                               
Albania                    5937.029526           0.0  
Austria                   36126.492700           1.0  
Belgium                   33692.605080           1.0  
Bosnia and Herzegovina     7446.298803           0.0  
Bulgaria                  10680.792820           0.0

## Groupby()


We often want to calculate summary statistics grouped by subsets or attributes within fields of our data. For example, we might want to calculate the average life expectancy in a particular year. Remember we've done this on a column using the describe() method. 

We can calculate basic statistics for all records in a single column using the syntax below:

In [266]:
data["lifeExp_2007"].describe()

count    142.000000
mean      67.007423
std       12.073021
min       39.613000
25%       57.160250
50%       71.935500
75%       76.413250
max       82.603000
Name: lifeExp_2007, dtype: float64

But if we want to summarize by one or more variables, for example continent, and then apply statistics. So we can use Pandas’ .groupby() method. 

Once we’ve created a groupby DataFrame, we can quickly calculate summary statistics by a group of our choice.

In [267]:
grouped_data = data.copy()
grouped_data = grouped_data.groupby("continent")

In [268]:
type(grouped_data)

pandas.core.groupby.generic.DataFrameGroupBy

We can now look at descriptive statistics for each of the columns in the original DataFrame, grouped by continent.

In [269]:
grouped_data.describe()

gdpPercap_1952                                            \
                   count          mean           std           min   
continent                                                            
Africa              52.0   1252.572466    982.952116    298.846212   
Americas            25.0   4079.062552   3001.727522   1397.717137   
Asia                33.0   5195.484004  18634.890865    331.000000   
Europe              30.0   5661.057435   3114.060493    973.533195   
Oceania              2.0  10298.085650    365.560078  10039.595640   

                                                                    \
                    25%           50%           75%            max   
continent                                                            
Africa       534.990554    987.025569   1454.886645    4725.295531   
Americas    2428.237769   3048.302900   3939.978789   13990.482080   
Asia         749.681655   1206.947913   3035.326002  108382.352900   
Europe      3241.132406   5142.469716   7236.794919   14734.232750   
Oceania    10168.840645  10298.085650  10427.330655   10556.575660   

          gdpPercap_1957                ...     pop_2002                \
                   count          mean  ...          75%           max   
continent                               ...                              
Africa              52.0   1385.236062  ...  16973552.75  1.199013e+08   
Americas            25.0   4616.043733  ...  26769436.00  2.876755e+08   
Asia                33.0   5787.732940  ...  66907826.00  1.280400e+09   
Europe              30.0   6963.012816  ...  20833960.25  8.235067e+07   
Oceania              2.0  11598.522455  ...  15637103.25  1.954679e+07   

          pop_2007                                                     \
             count          mean           std        min         25%   
continent                                                               
Africa        52.0  1.787576e+07  2.491773e+07   199579.0  2909226.50   
Americas      25.0  3.595485e+07  6.883378e+07  1056608.0  5675356.00   
Asia          33.0  1.155138e+08  2.896734e+08   708573.0  6426679.00   
Europe        30.0  1.953662e+07  2.362474e+07   301931.0  4780559.50   
Oceania        2.0  1.227497e+07  1.153885e+07  4115771.0  8195372.25   

                                                  
                  50%          75%           max  
continent                                         
Africa     10093310.5  19363654.50  1.350312e+08  
Americas    9319622.0  28674757.00  3.011399e+08  
Asia       24821286.0  69453570.00  1.318683e+09  
Europe      9493598.0  20849695.25  8.240100e+07  
Oceania    12274973.5  16354574.75  2.043418e+07  

[5 rows x 288 columns]

We can look at a specific statistic, mean, applied across the groupby DataFrame: 

In [240]:
grouped_data.mean()

gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
continent                                                                   
Africa        1252.572466     1385.236062     1598.078825     2050.363801   
Americas      4079.062552     4616.043733     4901.541870     5668.253496   
Asia          5195.484004     5787.732940     5729.369625     5971.173374   
Europe        5661.057435     6963.012816     8365.486814    10143.823757   
Oceania      10298.085650    11598.522455    12696.452430    14495.021790   

           gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
continent                                                                   
Africa        2339.615674     2585.938508     2481.592960     2282.668991   
Americas      6491.334139     7352.007126     7506.737088     7793.400261   
Asia          8187.468699     7791.314020     7434.135157     7608.226508   
Europe       12479.575246    14283.979110    15617.896551    17214.310727   
Oceania      16417.333380    17283.957605    18554.709840    20448.040160   

           gdpPercap_1992  gdpPercap_1997  ...      pop_1962      pop_1967  \
continent                                  ...                               
Africa        2281.810333     2378.759555  ...  5.702247e+06  6.447875e+06   
Americas      8044.934406     8889.300863  ...  1.733081e+07  1.922986e+07   
Asia          8639.690248     9834.093295  ...  5.140476e+07  5.774736e+07   
Europe       17061.568084    19076.781802  ...  1.534517e+07  1.603930e+07   
Oceania      20894.045885    24024.175170  ...  6.641759e+06  7.300207e+06   

               pop_1972      pop_1977      pop_1982      pop_1987  \
continent                                                           
Africa     7.305376e+06  8.328097e+06  9.602857e+06  1.105450e+07   
Americas   2.117537e+07  2.312271e+07  2.521164e+07  2.731016e+07   
Asia       6.518098e+07  7.225799e+07  7.909502e+07  8.700669e+07   
Europe     1.668784e+07  1.723882e+07  1.770890e+07  1.810314e+07   
Oceania    8.053050e+06  8.619500e+06  9.197425e+06  9.787208e+06   

               pop_1992      pop_1997      pop_2002      pop_2007  
continent                                                          
Africa     1.267464e+07  1.430448e+07  1.603315e+07  1.787576e+07  
Americas   2.957096e+07  3.187602e+07  3.399091e+07  3.595485e+07  
Asia       9.494825e+07  1.025238e+08  1.091455e+08  1.155138e+08  
Europe     1.860476e+07  1.896480e+07  1.927413e+07  1.953662e+07  
Oceania    1.045983e+07  1.112072e+07  1.172741e+07  1.227497e+07  

[5 rows x 36 columns]

Now let's look at a specific column. 

In [241]:
grouped_data["lifeExp_2007"].describe()

count       mean       std     min       25%      50%       75%  \
continent                                                                    
Africa      52.0  54.806038  9.630781  39.613  47.83400  52.9265  59.44425   
Americas    25.0  73.608120  4.440948  60.916  71.75200  72.8990  76.38400   
Asia        33.0  70.728485  7.963724  43.828  65.48300  72.3960  75.63500   
Europe      30.0  77.648600  2.979813  71.777  75.02975  78.6085  79.81225   
Oceania      2.0  80.719500  0.729027  80.204  80.46175  80.7195  80.97725   

              max  
continent          
Africa     76.442  
Americas   80.653  
Asia       82.603  
Europe     81.757  
Oceania    81.235

### Backup content on Transforms. 

In [13]:
data2 = data.T

In [14]:
data2.head()

0        1           2         3        4   \
country         Afghanistan  Bahrain  Bangladesh  Cambodia    China   
gdpPercap_1952      779.445  9867.08     684.244   368.469  400.449   
gdpPercap_1957      820.853  11635.8     661.637   434.038  575.987   
gdpPercap_1962      853.101  12753.3     686.342   496.914  487.674   
gdpPercap_1967      836.197  14804.7     721.186   523.432  612.706   

                             5        6          7        8        9   ...  \
country         Hong Kong China    India  Indonesia     Iran     Iraq  ...   
gdpPercap_1952          3054.42  546.566    749.682  3035.33  4129.77  ...   
gdpPercap_1957          3629.08  590.062      858.9  3290.26  6229.33  ...   
gdpPercap_1962          4692.65  658.347     849.29  4187.33  8341.74  ...   
gdpPercap_1967          6197.96  700.771    762.432  5906.73  8931.46  ...   

                         23            24         25         26       27  \
country         Philippines  Saudi Arabia  Singapore  Sri Lanka    Syria   
gdpPercap_1952      1272.88       6459.55    2315.14    1083.53  1643.49   
gdpPercap_1957      1547.94       8157.59     2843.1    1072.55  2117.23   
gdpPercap_1962      1649.55       11626.4    3674.74    1074.47  2193.04   
gdpPercap_1967      1814.13         16903    4977.42    1135.51  1881.92   

                     28        29       30                  31          32  
country          Taiwan  Thailand  Vietnam  West Bank and Gaza  Yemen Rep.  
gdpPercap_1952  1206.95   757.797  605.066             1515.59     781.718  
gdpPercap_1957  1507.86   793.577  676.285             1827.07      804.83  
gdpPercap_1962  1822.88    1002.2  772.049             2198.96     825.623  
gdpPercap_1967  2643.86   1295.46  637.123             2649.72     862.442  

[5 rows x 33 columns]